### ISSUE TITLE
Pupil Capture crashes on mac

### ISSUE THREAD

**Issue by tbalsdon** [pupil-labs/pupil#1155]
I am trying to implement eye-tracking and pupil dilation measurements with Pupil Labs for a visual psychophysics experiment. The capture app (latest) continually crashes and I don’t know what to do to try to fix it next!

Crashes occur on Mac OS 10.13.4, 10.12, 10.11, and I couldn’t even open the app on 10.9 (these are all different computers). I initially thought the crashes occurred because I was overloading the system running both matlab (to run the experiment) and pupil capture. I now have matlab running on a mac mini (late 2014, OS 10.13.4), it calls python functions (I couldn’t get a matlab zeromq binding to take) to chat with pupil on a separate mac mini (connected via ethernet cable) running pupil capture. The commands I send are fairly simple at the moment – I set the time, start recording, stop recording, query the time, and send some annotations. I make 6 recordings (blocks) of less than 10min each per session. I record the world camera at default settings, and I record from one eye at 120Hz, resolution 400x400 (I have also tried 192x192) (but do not save the video). I switched to 2d mapping mode, as this seemed to make things smoother (sometimes).

The first session with this set up worked almost perfectly (I lost ~4 frames of data one block and ~2 frames another, but I am ok with that). Since then it has become progressively worse. The next session I lost the second half of the last block (Capture didn’t crash, but the data was not being ‘captured’). The session after, capture began to crash, and I could not restart anything. The initial error that seems to set things off is:

video_capture.uvc_backend: capture failed to provide frames. Attempt to reinit.

It calls this error for both the eye and world cameras. It then reports calibration issues, then repeats this a bit. It then turns to:

uvc: could not set value… (‘Absolute exposure time’, ‘backlight compensation’)
uvc: could not init… (everything)
uvc: could not get Value. Must be read disabled

I have tried:
Switching USB ports (and ensuring it has its own bus – to confirm, this mac mini has USB3 ports).
Starting pupil capture from terminal.
Running pupil capture in default settings.
Deleting the recordings and the pupil capture folder (someone recommended this on a previous post).

On other set ups I have also tried reducing the frame rates / resolutions.

At one point I noticed one of the cameras was not being recognised in the USB port. I’m not sure why, or if this is causing the problems.

Similar crashes occur on other testing laptops (also macs, of the OSs listed above), even when only pupil capture is running. Sometimes the cameras are not recognised (could not init, failed to provide frames, et cetera), sometimes the application simply freezes and becomes unresponsive.

Any suggestions?

**Comment by papr** [comment]
Hey @tbalsdon 
I would like to see if I understood the situation correctly: Previously you did not have any problems and everything worked fine. Now the cameras are not detected by the system any more. Is this correct so far?

What do you mean by "crash"? Do the windows disappear and is the process terminated? Or does the camera video just not show up with in the World window?

Generally, it is an hardware issue if the cameras are not detected by the system. Could you make sure that the usb cable is connected correctly to the headset?

**Comment by tbalsdon** [comment]
Hi @papr 

>> Previously you did not have any problems and everything worked fine. Now the cameras are not detected by the system any more. Is this correct so far?

We have been having ongoing problems. I set up the device and software on a separate computer in an attempt to solve these problems. The first test it worked perfectly on this new set up, but then it returned to having the same problems as before. Basically, it temperamental, as opposed to completely broken.

>> What do you mean by "crash"? Do the windows disappear and is the process terminated? Or does the camera video just not show up with in the World window?

Most commonly, the application freezes - the video streams from the cameras are no longer updated and the buttons within the app are unresponsive. The application must be 'force quit' and restarted. Sometimes only the eye camera will freeze, whilst the world camera continues. Sometimes the application replaces the camera feeds with 'dummy' cameras (a grey screen).

>> Generally, it is an hardware issue if the cameras are not detected by the system. Could you make sure that the usb cable is connected correctly to the headset?

Do you mean at the USB-c to USB-3 point, or within the device? The device was purchased pre-built, I'm not sure I want to take it apart.

Thanks for your help!

**Comment by papr** [comment]
There is no reason to take the device apart! I meant the connection where the USB-c cable is connected to the device.

Do you run from source or do you run the bundled application?

**Comment by tbalsdon** [comment]
Sure! This is from a session I just tried to record.
[capture.log](https://github.com/pupil-labs/pupil/files/1914709/capture.log)

**Comment by papr** [comment]
The crash log does not show any indication for a crash. Did you experience a crash nonetheless?

**Comment by tbalsdon** [comment]
Not this time, only the problems listed in the log. Do you want me to run
it until it crashes?

**Comment by papr** [comment]
Do not worry too much about the uvc warnings as long as the video stream and the pupil detection work.

Yes, I will need you to reproduce the crash if possible and upload its log file. Else I am not able to fix the underlying issue.

**Comment by tbalsdon** [comment]
The uvc warnings are accompanied by data going unrecorded. The calibrations are also lost. It appears as if the app loses the cameras and then goes about finding them again.

I was able to reproduce the crash. The app became unresponsive and I had to force quit. Nothing additional appeared in the log file. In the terminal, the crash, and small pauses where the app doesn't appear responsive for a second or so, are accompanied by:

Estimated / selected altsetting bandwidth 392 / 800
!!!!Packets per transfer = 32 frameInterval = [a number I can't remember]

This has been linked to USB problems here: https://groups.google.com/forum/#!topic/pupil-discuss/s9RAqXSsoEE

At one point I checked whether the cameras were being recognised at the USB port. Two of them were missing. When I checked again a second later, they were all there.
